src/snapshot_derive/src/descriptor.rs

```diff
@@ -0,0 +1,428 @@
+use enum_field::*;
+use quote::{format_ident, quote};
+use struct_field::*;
+use union_field::*;
+
+/// An interface for generating serialzer and deserializers based on
+/// field descriptions.
+pub trait Descriptor {
+    /// Returns the serializer code block as a token stream.
+    fn generate_serializer(&self) -> proc_macro2::TokenStream;
+    /// Returns the deserializer code block as a token stream.
+    fn generate_deserializer(&self) -> proc_macro2::TokenStream;
+    /// Returns the curent version.
+    fn version(&self) -> u16;
+    /// Returns the type name as string.
+    fn ty(&self) -> String;
+}
+
+/// Describes a structure and it's fields.
+pub(crate) struct StructDescriptor {
+    // The structure type identifier.
+    pub ty: syn::Ident,
+    pub version: u16,
+    fields: Vec<StructField>,
+}
+
+/// Describes an enum and it's fields.
+pub(crate) struct EnumDescriptor {
+    pub ty: syn::Ident,
+    pub version: u16,
+    fields: Vec<EnumVariant>,
+}
+
+/// Describes an union and it's fields.
+pub(crate) struct UnionDescriptor {
+    pub ty: syn::Ident,
+    pub version: u16,
+    fields: Vec<UnionField>,
+}
+
+impl Descriptor for StructDescriptor {
+    fn generate_serializer(&self) -> proc_macro2::TokenStream {
+        let mut versioned_serializers = proc_macro2::TokenStream::new();
+
+        for i in 1..=self.version {
+            let mut versioned_serializer = proc_macro2::TokenStream::new();
+            let mut semantic_serializer = proc_macro2::TokenStream::new();
+
+            // Generate field and semantic serializers for all fields.
+            // Not all fields have semantic serializers defined and some fields
+            // might be missing in version `i`. In these cases the generate_serializer() and
+            // generate_semantic_serializer() will return an empty token stream.
+            for field in &self.fields {
+                versioned_serializer.extend(field.generate_serializer(i));
+                semantic_serializer.extend(field.generate_semantic_serializer(i));
+            }
+
+            // Generate the match arm for version `i`.
+            versioned_serializers.extend(quote! {
+                #i => {
+                    #semantic_serializer
+                    #versioned_serializer
+                }
+            });
+        }
+
+        // Generate code to map the app version to struct version and wrap the
+        // serializers with the `version` match.
+        let result = quote! {
+            // Get the struct version in target application version (app_version).
+            let version = version_map.get_type_version(app_version, &Self::name());
+            // We will use this copy to perform semantic serialization.
+            let mut copy_of_self = self.clone();
+            match version {
+                #versioned_serializers
+                _ => panic!("Unknown {} version {}.", &Self::name(), version)
+            }
+        };
+        result
```

Comment:
Part of this logic seems to be duplicated in all the structs that implement `Descriptor`. At least the last part is completely identical. I would move at leas part of this code to the trait:

```
pub trait Descriptor {
    fn generate_versioned_serializer(&self, version: u16) -> proc_macro2::TokenStream;

    fn generate_serializer(&self) -> proc_macro2::TokenStream {
        let mut versioned_serializers = proc_macro2::TokenStream::new();

        for i in 1..=self.version() {
            let versioned_serializer = self.generate_versioned_serializer(i);

            versioned_serializers.extend(quote! {
                #i => {
                    #versioned_serializer
                }
            });
        }

        let result = quote! {
            // Get the struct version for the input app_version.
            let version = version_map.get_type_version(app_version, &Self::name());
            // We will use this copy to perform semantic serialization.
            let mut copy_of_self = self.clone();
            match version {
                #versioned_serializers
                _ => panic!("Unknown {} version {}.", &Self::name(), version)
            }
        };

        result
    }
    /// Returns the deserializer code block as a token stream.
    fn generate_deserializer(&self) -> proc_macro2::TokenStream;
    /// Returns the curent version.
    fn version(&self) -> u16;
    /// Returns the type name as string.
    fn ty(&self) -> String;
}
```

Struct:
```
    fn generate_versioned_serializer(&self, i: u16) -> proc_macro2::TokenStream {
        let mut versioned_serializer = proc_macro2::TokenStream::new();
        let mut semantic_serializer = proc_macro2::TokenStream::new();

        // Generate field and semantic serializers for all fields.
        // Not all fields have semantic serializers defined and some fields
        // might be missing in version `i`. In these cases the generate_serializer() and
        // generate_semantic_serializer() will return an empty token stream.
        for field in &self.fields {
            versioned_serializer.extend(field.generate_serializer(i));
            semantic_serializer.extend(field.generate_semantic_serializer(i));
        }

        quote! {
            #semantic_serializer
            #versioned_serializer
        }
    }
```

Enum:
```
    fn generate_versioned_serializer(&self, i: u16) -> proc_macro2::TokenStream {
        let mut versioned_serializer = proc_macro2::TokenStream::new();

        for field in &self.fields {
            versioned_serializer.extend(field.generate_serializer(i));
        }

        quote! {
            match self {
                #versioned_serializer
            }
        }
    }
```

Union: 
```
    fn generate_versioned_serializer(&self, i: u16) -> proc_macro2::TokenStream {
        self.generate_union_serializer(i);
    }
```

If we create a trait for `StructField`, `EnumField`, `UnionField` that expose methods like `serialize`, `deserialize`, we can move even a bigger part of this logic to the `Descriptor` trait.

Probably the same comment is applicable for the deserialization part.

Reply:
This is a good suggestion, I will take it into consideration. The only concern I have is that if we keep on doing this we are polluting the Descriptor trait with many flavors of generate_* which would otherwise be confusing.

Reply:
I'll update the PR with another solution that IMO is much simpler.

Reply:
done